### PR TITLE
Draft: Only read stdout when running with `--more-verbose`

### LINKED ISSUE #4: Out of memory error with executables producing a lot of output
On my laptop with 16GB ram, the following command gets killed by the OOM killer. I believe `quickbench` is using too much memory from strictly reading the output of `hledger register` (about 700MB when piped to a file). I also notice `quickbench` is using considerable CPU time itself, which might skew timing results.

```
$ cd hledger
$ quickbench -f bench100k.sh 
Running 3 tests 1 times at 2025-02-27 13:15:14 CET:
Killed
$ echo $?
137
```

```
$ quickbench --help | head -n1
quickbench 1.0

$ hledger --version
hledger 1.30.1, linux-x86_64
```

## Cause

I believe the cause is the use of `readProcessWithExitCode`. From the [docs](https://hackage.haskell.org/package/process-1.6.25.0/docs/System-Process.html) for `readProcess`:

> `readProcess` forks an external process, reads its standard output strictly, blocking until the process terminates, and returns the output string.

Merge request incoming.

### PR DESCRIPTION
**Pull request by thomie**
When not running with `-V` or `--more-verbose`, we now pipe the stdout of the executable under test to `/dev/null`. This prevents quickbench from running out of memory in case the output is huge (GBs).

MINOR REMARK

With my version of GHC (9.6.6), all exceptions unfortunately get annoted `withBinaryFile`, see https://gitlab.haskell.org/ghc/ghc/-/issues/20886.

For example, when running `quickbench -w doesnotexist`, the error message is:

```
/dev/null: withBinaryFile: does not exist (No such file or directory)
```

When running `quickbench -w doesnotexist --more-verbose`, avoiding the call to `withBinaryFile`, the error message is the much clearer:

```
doesnotexist: readCreateProcess: posix_spawnp: does not exist (No such file or directory)
```

This is not ideal, but I believe using the latest version of GHC will fix it.

Fixes #4 .

### CODE REVIEW

File: src/QuickBench.hs [hunk 1]
```diff
@@ -198,17 +200,31 @@ replaceExecutable exe cmd = (unwords (exe:args), exe, args) where args = drop 1
 time :: Opts -> String -> [String] -> IO Float
 time opts exe args = do
   t1 <- getCurrentTime
-  (c, o, e) <- readProcessWithExitCode' exe args ""
+  maybeOutput <- if (moreVerbose opts)
+    then Just <$> readCreateProcess (proc exe args) ""
+    else const Nothing <$> callProcessIgnoreOutput exe args
   t2 <- getCurrentTime
-  when (not $ null o) $ outvv opts $ (if verbose opts then "\n" else "") ++ o
-  unless (c == ExitSuccess) $ out opts $ " (error: " ++ clean e ++ ") "
```
Comment (thomie) [inline src/QuickBench.hs]: Instead of capturing `stderr`, I now let the child process inherit `stderr` from the parent process (the default). So any error messages written to `stderr` by say hledger will be verbatim written to the `stderr` of quickbench.

Comment (thomie) [reply src/QuickBench.hs]: The other change is that now, a failure of the childprocess will immediately result in failure of the parent process. I consider this an improvement, because before errors from the child process could get obscured by the normal output table. But perhaps you want to keep the original behavior @simonmichael ?

**Comment by simonmichael** [comment]
Let's rebase this and I'll test.

**Comment by simonmichael** [comment]
> With my version of GHC (9.6.6), all exceptions unfortunately get annoted withBinaryFile

Thanks for noting this. I think we'll ignore it.

I see this change in output (old and new are compiled with newer GHC):

old:
```
$ quickbench -w nosuchexe echo
Running 1 tests 1 times with 1 executables at 2025-04-09 09:26:36 HST:
 (error: nosuchexe: readCreateProcessWithExitCode: posix_spawnp: does not exist (No such file or directory)) 
Best times:
+-----------++-----------+
|           || nosuchexe |
+===========++===========+
| nosuchexe ||      0.00 |
+-----------++-----------+
```

new:
```
$ stack exec -- quickbench -w nosuchexe echo
Running 1 tests 1 times with 1 executables at 2025-04-09 09:26:41 HST:
nosuchexe: createProcess: posix_spawnp: does not exist (No such file or directory)
```

The new behaviour seems better (reports the command line error once and exits early).

File: src/QuickBench.hs [hunk 2]
```diff
@@ -219,17 +221,31 @@ time :: Opts -> String -> [String] -> IO Float
...
   return $ realToFrac $ diffUTCTime t2 t1
 
--- ^ This variant also returns a failure when the executable is missing.
-readProcessWithExitCode' :: FilePath -> [String] -> String -> IO (ExitCode, String, String)
-readProcessWithExitCode' exe args inp =
-  readProcessWithExitCode exe args inp
-    `catch` \(e :: IOException) -> return (ExitFailure 1, "", show e)
+callProcessIgnoreOutput :: FilePath -> [String] -> IO ()
+callProcessIgnoreOutput cmd args =
+  withBinaryFile "/dev/null" WriteMode $ \devNull ->
```
Comment (simonmichael) [inline src/QuickBench.hs:235]: I think this is a unixism, unfortunately ? quickbench should be robust on Windows too.

File: src/QuickBench.hs (hunk 1 above)
Comment (simonmichael) [reply src/QuickBench.hs]: Hmm. Should quickbench require that all tested commands succeed ? I think it's better to continue if some fail, because sometimes you're benchmarking many things and partial information is better than none.

I think it would be an improvement if it could show the failed status (or some of the error output, if it's not too disruptive of layout) in the results table though ?

File: src/QuickBench.hs (hunk 2 above)
Comment (thomie) [reply src/QuickBench.hs:235]: Just to be sure: supporting only [WSL](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) is not sufficient?

Comment (simonmichael) [reply src/QuickBench.hs:235]: Yes, supporting only WSL is not sufficient.
